subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/ManifestInputStream.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.java.archives.internal;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * Reads manifest file, and automatically removes value wraps.
+ * Newlines are normalized to LF (for simplified processing later).
+ */
+public class ManifestInputStream extends InputStream {
```

Comment:
❓ I assume these stream classes were extracted from the other classes. I can see how it improve testability and readability. However, I'm concerned by the additional complexity of, what appears to be, handwritten manifest parser. Is there an implementation that already exists that we could rely on?

Reply:
I do not know if such an implementation exists.

Java's implementation is broken due to the use of `HashMap`.
Ant Manifest class doesn't support split multi-bytes characters.

I agree it is sad to have yet another manifest parser, however, the resulting implementation seems to be more-or-less clean.

I don't think it would be way better if I publish manifest parser under `com.github.vlsi:manifest-parser` coordinates. Would it?